src/java.base/share/classes/java/lang/ThreadLocal.java

```diff
@@ -155,21 +155,36 @@ public ThreadLocal() {
      * thread-local variable.  If the variable has no value for the
      * current thread, it is first initialized to the value returned
      * by an invocation of the {@link #initialValue} method.
+     * If the current thread does not support thread locals then
+     * this method returns its {@link #initialValue} (or {@code null}
+     * if the {@code initialValue} method is not overridden).
      *
      * @return the current thread's value of this thread-local
+     * @see Thread.Builder#allowSetThreadLocals(boolean)
      */
     public T get() {
-        Thread t = Thread.currentThread();
+        return get(Thread.currentThread());
+    }
+
+    /**
+     * Returns the value in the current carrier thread's copy of this
+     * thread-local variable.
+     */
+    T getCarrierThreadLocal() {
+        return get(Thread.currentCarrierThread());
+    }
+
+    private T get(Thread t) {
         ThreadLocalMap map = getMap(t);
-        if (map != null) {
+        if (map != null && map != ThreadLocalMap.NOT_SUPPORTED) {
```

Comment:
Due to the way `setInitialValue` is implemented, `getMap(t)` will currently be called twice when `ThreadLocal`s are disabled.

--------------------------------------------------------------------------------

This method should probably be changed so that when `map == ThreadLocalMap.NOT_SUPPORTED`, it simply does:
```java
return initialValue();
```

--------------------------------------------------------------------------------

```suggestion
        if (map != null) {
            if (map == ThreadLocalMap.NOT_SUPPORTED) {
                return initialValue();
            }
```

Reply:
It's benign but what you suggest may be clearer - thanks!